ts/sagas/messages.ts

```diff
@@ -0,0 +1,222 @@
+import {
+  BackendClient,
+  BackendClientT,
+  GetMessageT,
+  GetServiceT
+} from "../api/backend";
+import { apiUrlPrefix } from "../config";
+import {
+  MESSAGES_LOAD_CANCEL,
+  MESSAGES_LOAD_REQUEST
+} from "../store/actions/constants";
+import {
+  loadMessagesCancel,
+  loadMessagesFailure,
+  loadMessagesSuccess,
+  loadMessageSuccess,
+  MessagesLoadCancel,
+  MessagesLoadRequest
+} from "../store/actions/messages";
+import { loadServiceSuccess } from "../store/actions/services";
+import { messagesByIdSelectors } from "../store/reducers/entities/messages/messagesById";
+import { servicesByIdSelector } from "../store/reducers/entities/services/servicesById";
+import { sessionTokenSelector } from "../store/reducers/session";
+
+// An object containing MessageWithContent keyed by id
+export interface MessagesListObject {
+  [key: string]: MessageWithContent;
+}
+
+// An object containing ServicePublic keyed by id
+export interface ServicesListObject {
+  [key: string]: ServicePublic;
+}
+
+// An array of messages id
+export type MessagesIdsArray = ReadonlyArray<string>;
+
+// An array of services id
+export type ServicesIdsArray = ReadonlyArray<string>;
+
+/**
+ * A generator to load the message detail from the Backend
+ *
+ * @param {function} getMessage - The function that makes the Backend request
+ * @param {string} id - The id of the message to load
+ * @returns {(Error|MessageWithContent)}
+ */
+export function* loadMessage(
+  getMessage: TypeofApiCall<GetMessageT>,
+  id: string
+): IterableIterator<Effect> {
+  const response:
+    | BasicResponseType<MessageWithContent>
+    | undefined = yield call(getMessage, { id });
+
+  if (!response || response.status !== 200) {
+    return response ? response.value : Error();
+  } else {
+    // Send and action to store the new message in the store
+    yield put(loadMessageSuccess(response.value));
+    return response.value;
+  }
+}
+
+/**
+ * A generator to load the service details from the Backend
+ *
+ * @param {function} getService - The function that makes the Backend request
+ * @param {string} id - The id of the service to load
+ * @returns {(Error|ServicePublic)}
+ */
+export function* loadService(
+  getService: TypeofApiCall<GetServiceT>,
+  id: string
+): IterableIterator<Effect> {
+  const response: BasicResponseType<ServicePublic> | undefined = yield call(
+    getService,
+    { id }
+  );
+
+  if (!response || response.status !== 200) {
+    return response ? response.value : Error();
+  } else {
+    // Send and action to store the new service in the store
+    yield put(loadServiceSuccess(response.value));
+    return response.value;
+  }
+}
+
+/**
+ * A generator to load messages from the Backend.
+ * The messages returned by the Backend are filtered so the application downloads
+ * only the details of the messages and services not already in the redux store.
+ */
+export function* loadMessages(
+  backendClient: BackendClientT
+): IterableIterator<Effect> {
+  try {
```

Comment:
who is supposed to throw an exception here?

Reply:
In loadMessagesWatcher `yield cancel(lastTask.value);`

Reply:
I mean, here, in this code block, what method call is supposed to throw an exception?

Reply:
the execution flow is not very clear, it deserves an explanation

Reply:
thanks, pls add a comment with a link to that page